Add Knative Documentation + Reorganize example folders (#308)
vmware-samples/vcenter-event-broker-appliance: docs/kb/intro-functions.md

## Table of Contents
- [Knative](#knative)
  - [Knative Naming and Version Control](#knative-naming-and-version-control)
    - [Knative Service](#knative-service)
    - [Knative Trigger](#knative-service)
    - [Knative Combined Service and Trigger](#knative-combined-service-and-trigger)
    - [Knative Secrets](#knative-secrets)
    - [Knative Environment Variables](#knative-environment-variables)
- [OpenFaaS](#openfaas)
  - [OpenFaaS Naming and Version Control](#openfaas-naming-and-version-control)

## Knative

Users who directly want to jump into VMware vSphere-related function code might want to check out the examples we provide [here](/examples/knative).

### Knative Naming and Version Control

When it comes to authoring functions, it's important to understand the different components that make up a Knative function deployment. Let's take the following excerpt as an example:

### Knative Service

A Knative Service `kn-service.yaml` defines the container image that will be executed upon invocation.

```
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
 name: kn-ps-echo
 namespace: vmware-functions
spec:
 template:
  spec:
   containers:

>     - image: projects.registry.vmware.com/veba/kn-ps-echo:latest

Reviewer: Is this pushed?

Author: Yes

```

`kn-ps-echo`: The name of the Knative Service.

The value of this field:

- must not conflict with an existing function
- should not contain special characters, e.g. "$" or "/"
- should represent the intent of the function, e.g. "tag" or "tagging"

`vmware-functions`: The Kubernetes namespace to deploy all functions to. By default, this is `vmware-functions` which is automatically created as part of the VMware Event Broker Appliance setup.

> > **Note**: It is recommended to deploy all functions to the default `vmware-functions` namespace since the VMware Event Broker Appliance UI is automatically been configured to manage all functions and secrets within this namespace.

Reviewer: Good note! Might want to add that the broker lives there and would otherwise break fns

Author: 👍

`image:` The name of the resulting container image following Docker naming conventions `"<repo>/<image>:<tag>"`.

The value of this field:

- must resolve to a valid Docker container name (see convention above)
- should reflect the name of the function for clarity

> - should use a tag other than `"latest"`, e.g. `":0.2"` or `":$GIT_COMMIT"`

Reviewer: Better change your tag above then ;)

Author: 👍

## Knative Combined Service and Trigger

To simplify Knative function deployment, we can also combine the multiple manifest files into a single file. In the example below, the `function.yaml` contains both the Knative Service and Trigger definition.

```
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
 name: kn-ps-echo
 namespace: vmware-functions
spec:
 template:
  spec:
   containers:
    - image: projects.registry.vmware.com/veba/kn-ps-echo:latest
---
apiVersion: eventing.knative.dev/v1
kind: Trigger
metadata:
  name: veba-ps-echo-trigger
  namespace: vmware-functions
spec:
  broker: default
  filter:
    attributes:
       type: com.vmware.event.router/event
       subject: VmPoweredOffEvent
  subscriber:
    ref:
      apiVersion: serving.knative.dev/v1
      kind: Service
      name: kn-ps-echo
      namespace: vmware-functions
```

To deploy the Knative Service/Trigger, just run:

```
kubectl apply -f function.yaml
```

To undeploy the Knative Service/Trigger, just run:

```
kubectl delete -f function.yaml
```

## Knative Secrets

> Knative functions also support secrets or sensitive information which are passed in as part of the function deployment. Within the function, you can then access the secrets using an environmental variable and the name that the secret reference was created with.

Reviewer: ```suggestion
Knative functions also support secrets or sensitive information which are passed in as part of the function deployment. Within the function, you can then access the secrets using an environment variable and the name that the secret reference was created with.
```

> A secrets file (without file extension) should be created that contains the sensitive values including the structure you wish to process within your function. For example, you can encode your secrets into a JSON structure which means you can use the language specific JSON parser to easily extract out specific values.

Reviewer: Since you use env KEY name when creating the secret, the file extension actually does not matter.

> In the example below, the file containing your secret is called `secret`, this can be named anything but make sure it does not contain a file extension.

Reviewer: See comment above

```
cat > secret <<EOF
{
  "SLACK_WEBHOOK_URL": "YOUR-WEBHOOK-URL"
}
EOF
```

> Next, we need to create the Kubernetes secret by using the `--from-file` option, the name of the Kubernetes secret and the name of the environmental variable `SLACK_SECRET` which will be accessed by your functional handler.

Reviewer: ```suggestion
Next, we need to create the Kubernetes secret by using the `--from-file` option, the name of the Kubernetes secret and the name of the environment variable `SLACK_SECRET` which will be accessed by your functional handler.
```

> > **Note**: The environmental variable name should begin with an uppercase letter and contain only uppercase letters(A-Z), numbers(0-9) and underscores(_)

Reviewer: See recommended changes on `environmental` above

> > **Note**: The environmental variable name should begin with an uppercase letter and contain only uppercase letters(A-Z), numbers(0-9) and underscores(_)

Reviewer: ```suggestion
> **Note**: The environment variable name must only contain uppercase letters(A-Z), numbers(0-9) and underscores(_)
```

```
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
 name: kn-ps-echo
 namespace: vmware-functions
spec:
 template:
  metadata:
  spec:
   containers:
    - image: projects.registry.vmware.com/veba/kn-ps-echo:latest
      envFrom:
        - secretRef:
            name: slack-secret
---
apiVersion: eventing.knative.dev/v1
kind: Trigger
metadata:
  name: veba-ps-echo-trigger
  namespace: vmware-functions
spec:
  broker: default
  filter:
    attributes:
       type: com.vmware.event.router/event
       subject: VmPoweredOffEvent
  subscriber:
    ref:
      apiVersion: serving.knative.dev/v1
      kind: Service
      name: kn-ps-echo
      namespace: vmware-functions
```

Finally, to access the secret from within your function handler, use the language specific option to access the environmental variable that you had named earlier called `SLACK_SECRET` and decode the contents.

Here is an example using the PowerShell language to access the environmental variable:

```
$jsonSecrets = ${env:SLACK_SECRET} | ConvertFrom-Json
```

> ## Knative Environment Variables

Reviewer: +1 for using "Environment" :)

Knative functions also support defining additional environmental variables that can be passed in as part of the function deployment. To do so, define a new `env:` section with a list of name and values. In the example below, the additional variable is named `FUNCTION_DEBUG` and contains the vaue of `"true"` (notice this must be an encapsulated string).

> > **Note**: The environmental variable name should begin with an uppercase letter and contain only uppercase letters(A-Z), numbers(0-9) and underscores(_)

Reviewer: ```suggestion
> **Note**: The environment variable name must only contain uppercase letters(A-Z), numbers(0-9) and underscores(_)
```
